### DISCUSSION TITLE
New collection functions for server side (CreateCollectionRule,UpdateCollectionRule,DeleteCollectionRule)

### DISCUSSION THREAD

**Discussion by mbos2** [appwrite/appwrite#1620] [Ideas]
Hello everyone.

I would like to propose new functions for collections on appwrite server side (node-appwrite).

### Recap of current update behaviour
This is updateCollection function with all parameters except `list`
```
updateCollection(
'collectionID,
'collectionName',
[readPermissions],
[writePermissions],
[arrayOfRuleObjects]
);
```
And for example
```
updateCollection(
'123456789,
'Test collection',
['*'],
['*'],
[
new CollectionRule('Name', 'name', 'text', '', true, false) // Existing rule
new CollectionRule('Age', 'age', 'text', '', true, false) // Existing rule
new CollectionRule('Adress', 'adress', 'text', '', true, false) // Existing rule
new CollectionRule('Gender', 'age', 'text', '', true, false), // NEW rule 
]
);
```
Currently, to update a collection that already has rules in place would require a user to insert currently existing rules along with new ones, like shown in the example, because function completelly rewrites a collection based on parameters.
- Same or similar example code applies when user wants to delete a rule.
- Same or similar example code applies when user wants to change a name of Collection as well.

### Use case

With current behaviour user needs to constantly repeat rules he wants to keep in the collection when updating it in any way. 
This makes user write a lot of repetitive code when writing **database** migrations for collection(s) or just updating collection in general. 

In case of larger collections, writing such migrations would require extra comments in order for others to know what has happened. Without comments it would take much longer to determine what has changed in latest migration in comparison to previous one, especially with large applications.

### Proposal

To avoid repetitive code I propose functions:
```
insertCollectionRule('collectionID', {newRuleObject});
updateCollectionRule('collectionID', 'collectionRuleKey | collectionRuleUniqueKey', {newRuleObject});
deleteCollectionRule('collectionID', 'collectionRuleKey | collectionRuleUniqueKey');
```

This would mean that it would only affect collection rules while keeping all other collection parameters the same.

Also, this might require for rules to have unique ID or at least unique `key` as currently we can have multiple indentical rules in our collection(s) for update/delete to target proper rule.

### Other areas of implementation

Such behaviour could also be considered for other collection parameters as well.

Appwrite currently has such functions for documents:
- https://appwrite.io/docs/server/database?sdk=nodejs-default#databaseUpdateDocument
- https://appwrite.io/docs/server/database?sdk=nodejs-default#databaseDeleteDocument

**Comment by kodumbeats** [comment]
Most of this proposal has been captured in the upcoming database refactor. We've deprecated `rules` in favor of `attributes`, which map one-to-one with SQL columns, rely on primitive types, and can allow for custom indexing. Check out this gist for more information:

https://gist.github.com/kodumbeats/d0f8d8305b28f1b3f3d52f0ada358551
